File: packages/api/src/trueppm_api/apps/projects/schema_migrations.py
"""Forward-migration registry for user-saved JSON state (ADR-0086 / ADR-0201).

Every user-saved JSON payload (saved views, filters, dashboards) carries a
top-level ``schema_version: int``. No business code reads a raw stored payload;
it dispatches through this registry, which upgrades a payload to the current
version **on read** before any consumer sees it.

The registry is a generic, surface-keyed chain of pure ``v(n) -> v(n+1)``
transforms. Reading a ``v1`` payload when the surface's current version is ``v3``
applies ``v1->v2`` then ``v2->v3``. Writes always emit the current version.

Design contract (mirrored in ``packages/web/src/lib/schemaMigrations.ts``):

* A stored payload with no ``schema_version`` key is treated as version ``0``.
* A payload at a version *newer* than the running code (e.g. after a downgrade)
  is a hard error, not a silent best-effort read.
* Surface keys are shared string constants, not duplicated literals, so the API
  and web registries can be audited against each other.

New saved-state surfaces register their current version and their upgrade steps
here (one ``register_migration`` call per ``v(n) -> v(n+1)`` transform) and read
through :func:`migrate_payload`.
"""

from __future__ import annotations

from collections.abc import Callable, Mapping
from typing import Any

# Surface keys — shared constants (mirror ``SURFACE_*`` in the web registry).
SURFACE_BOARD_SAVED_VIEW = "board_saved_view"

#: A single ``v(n) -> v(n+1)`` transform. Receives a payload known to be at
#: ``from_version`` and returns the payload at ``from_version + 1``. Must be pure.
Migration = Callable[[Mapping[str, Any]], dict[str, Any]]

# {surface: {from_version: transform}} — the ordered upgrade chain per surface.
_MIGRATIONS: dict[str, dict[int, Migration]] = {}

# {surface: current_version} — the version a fresh write emits and a read targets.
_CURRENT_VERSIONS: dict[str, int] = {}


class UnknownSchemaVersionError(ValueError):
    """A stored payload is at a version newer than the running code supports.

    Raised deliberately instead of best-effort reading a future-version payload
    (e.g. after a rollback to an older deployment). Loud failure is preferred to
    silently mis-reading data written by newer code (ADR-0086).
    """


def register_surface(surface: str, current_version: int) -> None:
    """Register a surface and its current schema version.

    Args:
        surface: Stable surface key (use a ``SURFACE_*`` constant).
        current_version: The version a fresh write emits; reads upgrade to it.
    """
    _CURRENT_VERSIONS[surface] = current_version
    _MIGRATIONS.setdefault(surface, {})


def register_migration(surface: str, from_version: int, fn: Migration) -> None:
    """Register a single ``from_version -> from_version + 1`` transform.

    Args:
        surface: Stable surface key (use a ``SURFACE_*`` constant).
        from_version: The version this transform upgrades *from*.
        fn: Pure function upgrading a ``from_version`` payload to the next version.
    """
    _MIGRATIONS.setdefault(surface, {})[from_version] = fn


def current_version(surface: str) -> int:
    """Return the current schema version for a surface (default 1 if unregistered)."""
    return _CURRENT_VERSIONS.get(surface, 1)


def stored_version(payload: Mapping[str, Any]) -> int:
    """Return the version stored in a payload, treating an absent key as 0.

    A missing ``schema_version`` means the payload predates the convention
    (ADR-0086), so it is treated as version 0 and run through the full chain.
    """
    raw = payload.get("schema_version", 0) if isinstance(payload, Mapping) else 0
    try:
        return int(raw)
    except (TypeError, ValueError):
        return 0


def migrate_payload(
    surface: str,
    payload: Mapping[str, Any],
    from_version: int | None = None,
) -> tuple[dict[str, Any], int]:
    """Upgrade a stored payload to the surface's current version on read.

    Applies the ordered ``v(n) -> v(n+1)`` chain until the payload reaches the
    current version, then stamps ``schema_version`` on the result. Re-running
    against an already-current payload is a no-op (empty chain), so this is
    idempotent.

    Args:
        surface: Stable surface key (use a ``SURFACE_*`` constant).
        payload: The stored JSON payload (may lack ``schema_version``).
        from_version: Override the detected version; defaults to the payload's
            own ``schema_version`` (absent => 0).

    Returns:
        A ``(upgraded_payload, current_version)`` tuple. ``upgraded_payload``
        carries ``schema_version`` set to the current version.

    Raises:
        UnknownSchemaVersionError: If the payload is at a version newer than the
            surface's registered current version.
    """
    target = current_version(surface)
    version = stored_version(payload) if from_version is None else from_version

    if version > target:
        raise UnknownSchemaVersionError(
            f"Payload for surface {surface!r} is at schema_version {version}, "
            f"but this code only supports up to {target}."
        )

    result: dict[str, Any] = dict(payload) if isinstance(payload, Mapping) else {}
    steps = _MIGRATIONS.get(surface, {})
    while version < target:
        transform = steps.get(version)
        if transform is None:
            # A gap in the chain is a programming error — a version bump landed
            # without its transform. Fail loudly rather than mis-read.
            raise UnknownSchemaVersionError(
                f"No migration registered for surface {surface!r} from version "
                f"{version} (current version is {target})."
            )
        result = transform(result)
        version += 1

    result["schema_version"] = target
    return result, target


# ---------------------------------------------------------------------------
# board_saved_view surface (#191, useBoardSavedViews)
# ---------------------------------------------------------------------------

# The six canonical config keys and their documented defaults. Kept in sync with
# BoardSavedViewSerializer.validate_config (the write-side normalizer).
_BOARD_VIEW_DEFAULTS: dict[str, Any] = {
    "sort": "priority",
    "show_wip": True,
    "show_col_tints": True,
    "evm_mode": "off",
    "show_cost": False,
    "risk_linked_only": False,
}


def _board_view_v0_to_v1(payload: Mapping[str, Any]) -> dict[str, Any]:
    """Backfill the six canonical config keys on a pre-convention board view.

    A version-0 payload predates ``schema_version`` and may be missing keys that
    were added to the config shape over time. Fill any absent canonical key with
    its documented default; keep existing values (including any extra keys, which
    the write-side validator drops on the next save).
    """
    upgraded = dict(payload)
    for key, default in _BOARD_VIEW_DEFAULTS.items():
        upgraded.setdefault(key, default)
    return upgraded


register_surface(SURFACE_BOARD_SAVED_VIEW, current_version=1)
register_migration(SURFACE_BOARD_SAVED_VIEW, 0, _board_view_v0_to_v1)

File: packages/web/src/lib/schemaMigrations.ts
/**
 * Forward-migration registry for user-saved JSON state (ADR-0086 / ADR-0204).
 *
 * Mirror of the API-side registry
 * (`packages/api/src/trueppm_api/apps/projects/schema_migrations.py`). Every
 * user-saved JSON payload (saved views, filters, dashboards) carries a
 * `schema_version`. No consumer reads a raw stored payload; it dispatches
 * through this registry, which upgrades a payload to the current version on read
 * before any business logic sees it.
 *
 * Contract (identical to the API side):
 * - A payload with no `schema_version` is treated as version 0.
 * - A payload at a version newer than this code supports is a hard error, not a
 *   silent best-effort read.
 * - Surface keys are shared string constants so the two registries stay auditable
 *   against each other.
 */

/** A single `v(n) -> v(n+1)` transform. Pure; receives a payload at `from`, returns it at `from + 1`. */
export type Migration = (payload: Record<string, unknown>) => Record<string, unknown>;

/** Stable surface keys — mirror `SURFACE_*` in the API registry. */
export const SURFACE_BOARD_SAVED_VIEW = 'board_saved_view';

interface SurfaceRegistration {
  currentVersion: number;
  /** Map of `from_version` -> transform to `from_version + 1`. */
  steps: Map<number, Migration>;
}

const registry = new Map<string, SurfaceRegistration>();

/** Register a surface and its current schema version. */
export function registerSurface(surface: string, currentVersion: number): void {
  const existing = registry.get(surface);
  if (existing) {
    existing.currentVersion = currentVersion;
  } else {
    registry.set(surface, { currentVersion, steps: new Map() });
  }
}

/** Register a single `fromVersion -> fromVersion + 1` transform for a surface. */
export function registerMigration(surface: string, fromVersion: number, fn: Migration): void {
  let entry = registry.get(surface);
  if (!entry) {
    entry = { currentVersion: 1, steps: new Map() };
    registry.set(surface, entry);
  }
  entry.steps.set(fromVersion, fn);
}

/** Current schema version for a surface (defaults to 1 if unregistered). */
export function currentVersion(surface: string): number {
  return registry.get(surface)?.currentVersion ?? 1;
}

/** Version stored in a payload, treating an absent/invalid key as 0. */
export function storedVersion(payload: Record<string, unknown> | null | undefined): number {
  const raw = payload?.['schema_version'];
  const n = typeof raw === 'number' ? raw : Number(raw);
  return Number.isFinite(n) ? Math.trunc(n) : 0;
}

/**
 * Upgrade a stored payload to the surface's current version on read.
 *
 * Applies the ordered `v(n) -> v(n+1)` chain until the payload is current, then
 * stamps `schema_version`. Re-running against an already-current payload is a
 * no-op, so this is idempotent.
 *
 * @throws Error if the payload is at a version newer than this code supports, or
 *   if a step is missing from the chain (a version bump landed without its
 *   transform).
 */
export function migratePayload(
  surface: string,
  payload: Record<string, unknown>,
  fromVersion?: number,
): { payload: Record<string, unknown>; version: number } {
  const target = currentVersion(surface);
  let version = fromVersion ?? storedVersion(payload);

  if (version > target) {
    throw new Error(
      `Payload for surface "${surface}" is at schema_version ${version}, but this code only supports up to ${target}.`,
    );
  }

  let result: Record<string, unknown> = { ...payload };
  const steps = registry.get(surface)?.steps ?? new Map<number, Migration>();
  while (version < target) {
    const transform = steps.get(version);
    if (!transform) {
      throw new Error(
        `No migration registered for surface "${surface}" from version ${version} (current version is ${target}).`,
      );
    }
    result = transform(result);
    version += 1;
  }

  result['schema_version'] = target;
  return { payload: result, version: target };
}

// ---------------------------------------------------------------------------
// board_saved_view surface (issue 191, useBoardSavedViews)
// ---------------------------------------------------------------------------

/** The six canonical config keys and their defaults — mirror of the API side. */
const BOARD_VIEW_DEFAULTS: Record<string, unknown> = {
  sort: 'priority',
  show_wip: true,
  show_col_tints: true,
  evm_mode: 'off',
  show_cost: false,
  risk_linked_only: false,
};

/** Backfill the canonical config keys on a pre-convention (v0) board view payload. */
function boardViewV0ToV1(payload: Record<string, unknown>): Record<string, unknown> {
  const upgraded = { ...payload };
  for (const [key, value] of Object.entries(BOARD_VIEW_DEFAULTS)) {
    if (upgraded[key] === undefined) {
      upgraded[key] = value;
    }
  }
  return upgraded;
}

/**
 * The three filter-facet keys added in v2 (issue 1918) — mirror of the API-side
 * `_BOARD_VIEW_FACET_DEFAULTS`. An empty list means "no constraint", matching
 * `boardFacets.EMPTY_FACETS`.
 */
const BOARD_VIEW_FACET_DEFAULTS: Record<string, unknown> = {
  filter_assignees: [],
  filter_priority: [],
  filter_due: [],
};

/** Backfill the filter-facet keys onto a pre-#1918 (v1) board view payload. */
function boardViewV1ToV2(payload: Record<string, unknown>): Record<string, unknown> {
  const upgraded = { ...payload };
  for (const [key, value] of Object.entries(BOARD_VIEW_FACET_DEFAULTS)) {
    if (upgraded[key] === undefined) {
      upgraded[key] = value;
    }
  }
  return upgraded;
}

registerSurface(SURFACE_BOARD_SAVED_VIEW, 2);
registerMigration(SURFACE_BOARD_SAVED_VIEW, 0, boardViewV0ToV1);
registerMigration(SURFACE_BOARD_SAVED_VIEW, 1, boardViewV1ToV2);
